dcos/cosmos.py

```diff
@@ -0,0 +1,401 @@
+                       (see py:func:`request`)
+        :type kwargs: dict
+        :return: the Response object returned by cosmos
+        :rtype: requests.Response
+        """
+        if not self._endpoint_exists(endpoint):
+            raise DCOSException(
+                'Cosmos called with incorrect endpoint {}'.format(endpoint))
+        url = self._get_endpoint_url(endpoint)
+        request_versions = self._get_request_version_preferences(endpoint)
+        headers_preference = list(map(
+            lambda version: self._get_header(
+                endpoint, version, headers),
+            request_versions))
+        http_request_type = self._get_http_method(endpoint)
+        return self._call_cosmos(
+            url,
+            http_request_type,
+            headers_preference,
+            data,
+            remove_nones(json),
+            **kwargs)
+
+    def _call_cosmos(self,
+                     url,
+                     http_request_type,
+                     headers_preference,
+                     data=None,
+                     json=None,
+                     **kwargs):
+        """
+        Gets a Response object obtained by calling cosmos
+        at the url 'url'. Will attempt each of the headers
+        in headers_preference in order until success.
+
+        :param url: the url of a cosmos endpoint
+        :type url: str
+        :param headers_preference: a list of request headers
+        in order of preference. Each header
+        will be attempted until they all fail or the request succeeds.
+        :type headers_preference: list[dict[str, str]]
+        :param data: the request's body
+        :type data: dict | bytes | file-like object
+        :param json: JSON request body
+        :type json: dict
+        :param kwargs: Additional arguments to requests.request
+                       (see py:func:`request`)
+        :type kwargs: dict
+        :return: response returned by calling cosmos at url
+        :rtype: requests.Response
+        """
+        try:
+            headers = headers_preference[0]
+            if http_request_type is 'post':
+                response = http.post(
+                    url, data=data, json=json, headers=headers, **kwargs)
+            else:
+                response = http.get(
+                    url, data=data, json=json, headers=headers, **kwargs)
+            if not _matches_expected_response_header(headers,
+                                                     response.headers):
+                raise DCOSException(
+                    'Server returned incorrect response type, '
+                    'expected {} but got {}'.format(
+                        headers.get('Accept'),
+                        response.headers.get('Content-Type')))
+        except DCOSBadRequest as e:
+            if len(headers_preference) > 1:
+                # reattempt with one less item in headers_preference
```

Comment:
This shouldn't be necessary unless theres a bug in cosmos with content negotiation